src/pending_connection/connect.rs

```diff
@@ -1,208 +1,261 @@
+                }
+                control_type => Err(HandshakeExpected(ShakeType::Induction, control_type)),
+            },
+            (ConclusionResponseWait(_), Packet::Control(control)) => match control.control_type {
+                ControlTypes::Handshake(shake) => self.wait_for_conclusion(from, shake),
+                control_type => Err(HandshakeExpected(ShakeType::Conclusion, control_type)),
+            },
+            (_, Packet::Data(data)) => Err(ControlExpected(ShakeType::Induction, data)),
+            (_, _) => Ok(None),
+        }
+    }
+
+    pub fn next_tick(&mut self, _now: Instant) -> ConnectResult {
+        match self.1.clone() {
+            Configured(init_seq_num) => self.on_start(init_seq_num),
+            InductionResponseWait(request_packet) => Ok(Some((request_packet, self.0.remote))),
+            ConclusionResponseWait(request_packet) => Ok(Some((request_packet, self.0.remote))),
+            _ => Ok(None),
         }
     }
 }
 
-async fn get_initial_handshake<T>(
+pub async fn connect<T>(
     sock: &mut T,
-    local_addr: IpAddr,
-    local_sockid: SocketID,
     remote: SocketAddr,
-) -> Result<(i32, HandshakeControlInfo), Error>
+    local_sockid: SocketID,
+    local_addr: IpAddr,
+    tsbpd_latency: Duration,
+    _crypto: Option<(u8, String)>,
+) -> Result<Connection, Error>
 where
     T: Stream<Item = Result<(Packet, SocketAddr), Error>>
         + Sink<(Packet, SocketAddr), Error = Error>
         + Unpin,
 {
-    let mut send_interval = interval(Duration::from_millis(100));
-
-    info!("Connecting to {}...", remote);
-
-    let request_packet = Packet::Control(ControlPacket {
-        dest_sockid: SocketID(0),
-        timestamp: 0, // TODO: this is not zero in the reference implementation
-        control_type: ControlTypes::Handshake(HandshakeControlInfo {
-            init_seq_num: rand::random(),
-            max_packet_size: 1500, // TODO: take as a parameter
-            max_flow_size: 8192,   // TODO: take as a parameter
-            socket_id: local_sockid,
-            shake_type: ShakeType::Induction,
-            peer_addr: local_addr,
-            syn_cookie: 0,
-            info: HandshakeVSInfo::V4(SocketType::Datagram),
-        }),
-    });
-
-    sock.send((request_packet.clone(), remote)).await?;
-
-    info!("Sent first packet to {}", remote);
-
-    // get first response packet
-    let (timestamp, hs_info) = loop {
-        // just drop the future that didn't finish first
-        let (packet, addr) = select! {
-            _ = send_interval.tick().fuse() => {sock.send((request_packet.clone(), remote)).await?; continue},
-            res = get_packet(sock).fuse() => res?
+    let mut connect = Connect(
+        ConnectConfiguration {
+            remote,
+            local_sockid,
+            local_addr,
+            tsbpd_latency,
+        },
+        ConnectState::new(),
+    );
+
+    let mut tick_interval = interval(Duration::from_millis(100));
+    loop {
```

Comment:
I like this design--it's quite explicit. 

One thing that could be solved at the same time is right now it would be trivial to design a DOS attack on this by just sending an induction packet over and over, which would inhibit anone else from connecting. I think the resolution would be to have something like `HashMap<SocketAddr, ConnectState>`, and just use the address that resolves first.

Reply:
Yeah. out of the scope of this PR though.